source/common/router/router.cc

```diff
@@ -778,5 +763,37 @@ ProdFilter::createRetryState(const RetryPolicy& policy, Http::HeaderMap& request
                                 priority);
 }
 
+void Filter::UpstreamRequest::setRequestEncoder(Http::StreamEncoder& request_encoder) {
+  request_encoder_ = &request_encoder;
+  // Now that there is an encoder, have the connection manager inform the manager when the
+  // downstream buffers are overrun.  This may result in immediate watermark callbacks referencing
+  // the encoder.
+  parent_.callbacks_->addDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+}
+
+void Filter::UpstreamRequest::clearRequestEncoder() {
+  // Before clearing the encoder, unsubscribe from callbacks.  This may result
+  // in some clean up calls to onBelowWriteBufferLowWatermark which reference
```

Comment:
For HTTP2 it's fine to consider the stream dead when it's reset and not update flow control - the H2 layer will take care of making sure all the stream data is consumed.

For HTTP/1 I assume we reuse connections and so we either need to read-enable in the router, or we need to re-enable somewhere else to make sure the next user of an upstream connection doesn't pick up a Network::Connection which won't be read from.